### PR TITLE
[flang][acc] Improve lowering of Fortran optional in data clause

### PR DESCRIPTION
**Pull request by razvanlupusoru**
Fortran optional arguments are effectively null references. To deal with this possibility, flang lowering of OpenACC data clauses creates three if-else regions when preparing the data pointer for the data clause:
1) Load box value from box reference
2) Load box addr from box value
3) Load box dims from box value

However, this pattern makes it more complicated to find the original box reference. Effectively, the first if-else region to get the box value is not needed - since the value can be loaded before the corresponding `fir.box_addr` and `fir.box_dims` operations. Thus, reduce the number of if-else regions by deferring the box load to the use sites.

For non-optional cases, the old functionality is left alone - which preloads the box value.

### CODE REVIEW

File: flang/lib/Lower/OpenACC.cpp [hunk 1]
```diff
@@ -108,6 +120,9 @@ createDataEntryOp(fir::FirOpBuilder &builder, mlir::Location loc,
               })
               .getResults()[0];
     } else {
+      if (fir::isBoxAddress(baseAddr.getType())) {
+        baseAddr = builder.create<fir::LoadOp>(loc, baseAddr);
+      }
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:125]:
```suggestion
      if (fir::isBoxAddress(baseAddr.getType()))
        baseAddr = builder.create<fir::LoadOp>(loc, baseAddr);
```

**Review by clementval — APPROVED** [review]
LGTM. 

Just some nits and looks like you have clang-format issues.

**Review by vzakhari — APPROVED** [review]
Looks great (modulo Valentin's comments)! Thank you, Razvan!

File: flang/lib/Lower/OpenACC.cpp (hunk 1 above)
Comment (razvanlupusoru) [reply flang/lib/Lower/OpenACC.cpp]: Done!

**Pull request merged**